File: LLM/LLM/Core/TransformerModel.cs
using Console = System.Console;
using Exception = System.Exception;
using TorchSharp;
using TorchSharp.Modules;
using static TorchSharp.torch;
using static TorchSharp.torch.nn; // Para ModuleList, Linear, Embedding, LayerNorm

namespace LLM.Core;

public class TransformerModel : Module
{
    private readonly Embedding tokenEmbedding;
    private readonly Embedding positionEmbedding; // Embedding posicional aprendível
    private readonly ModuleList<Module> transformerBlocks = new(); // Usar ModuleList genérico
    private readonly LayerNorm finalNorm; // Normalização final antes da saída
    private readonly Linear outputLinear; // Camada de saída para logits

    // Tornar maxSeqLen e vocabSize públicos para acesso pelo Trainer/Inferência
    public readonly int maxSeqLen;
    public readonly int embeddingDim;
    public readonly int vocabSize;
    public Device device;

    public TransformerModel(
        string name,
        int vocabSize,
        int maxSeqLen = 512,
        int embeddingDim = 256,
        int numHeads = 4,
        int numLayers = 2,
        double dropout = 0.1)
        : base(name)
    {
        device = cuda.is_available() ? CUDA : CPU;
        this.vocabSize = vocabSize;
        this.maxSeqLen = maxSeqLen;
        this.embeddingDim = embeddingDim;

        tokenEmbedding = Embedding(vocabSize, embeddingDim);
        positionEmbedding = Embedding(maxSeqLen, embeddingDim);

        // Instanciar usando o tipo genérico torch.nn.Module
        transformerBlocks = new ModuleList<torch.nn.Module>();
        for (int i = 0; i < numLayers; i++)
        {
            transformerBlocks.Append(new TransformerBlock($"Block_{i}", embeddingDim, numHeads, maxSeqLen, dropout));
        }

        finalNorm = LayerNorm(embeddingDim);
        outputLinear = Linear(embeddingDim, vocabSize);

        // Inicialização de pesos (opcional, mas recomendado para estabilidade)
        // foreach (var p in this.parameters())
        // {
        //      if (p.IsInitialized) // Check if parameter requires gradients and is not frozen
        //      {
        //          // Exemplo de inicialização Kaiming Uniforme (comum para ReLU/GELU)
        //          torch.nn.init.kaiming_uniform_(p, nonlinearity: torch.nn.init.Nonlinearity.ReLU);
        //          // Para camadas de normalização ou bias, pode ser diferente
        //      }
        // }


        RegisterComponents(); // Garante que módulos filhos são registrados e movidos para o device
    }

    public Tensor forward(Tensor input)
    {
        var (batchSize, seqLen) = (input.shape[0], input.shape[1]);

        // O forward do modelo espera uma sequência de tamanho ATÉ maxSeqLen.
        // Se a sequência de entrada for maior, precisamos TRUNCAR aqui.
        // O endpoint de Inferência agora passa a sequência gerada COMPLETA,
        // então o truncamento deve ocorrer DENTRO do forward do modelo.
        if (seqLen > maxSeqLen)
        {
            // Truncar a sequência de entrada para pegar APENAS os últimos maxSeqLen tokens
            long start = seqLen - maxSeqLen; // Calcula o índice inicial
            long end = seqLen;               // O índice final (exclusivo)
            long step = 1;                   // O passo

            // Slicing em TorchSharp: tensor.slice(dim, start, end, step)
            input = input.slice(1, start, end, step);

            seqLen = maxSeqLen; // Atualiza seqLen para o novo comprimento
            // Console.WriteLine($"Aviso: Sequência de entrada truncada para {maxSeqLen} tokens.");
        }
        // Se a sequência for menor que maxSeqLen, o embedding posicional
        // ainda usará posições de 0 a seqLen-1. Não precisa de padding aqui,
        // pois a máscara causal já lida com o triângulo inferior.

        var device = input.device; // Obter o dispositivo do tensor de entrada

        // Criar tensor de posições dinamicamente NO MESMO DISPOSITIVO DO INPUT
        var positions = torch.arange(0, seqLen, dtype: ScalarType.Int64, device: device)
            .unsqueeze(0) // [1, seqLen]
            .expand(batchSize, seqLen); // [batch, seqLen]

        // Calcular embeddings + posições
        var tokenEmbeds = tokenEmbedding.forward(input); // [batch, seqLen, embedDim]
        var posEmbeds = positionEmbedding.forward(positions); // [batch, seqLen, embedDim]
        var x = tokenEmbeds + posEmbeds; // [batch, seqLen, embedDim]

        // Aplicar dropout no embedding (comum)
        // Use this.training para aplicar dropout apenas durante o treino
        // x = torch.nn.functional.dropout(x, p: dropout, training: this.training); // Use o 'dropout' configurado no construtor

        // Passar pelos blocos Transformer
        foreach (var block in transformerBlocks)
        {
            dynamic dynamicBlock = block;
            x = dynamicBlock.forward(x);
        }

        // Normalização final
        x = finalNorm.forward(x);

        // Camada linear final para obter logits sobre o vocabulário
        var logits = outputLinear.forward(x); // [batch, seqLen, vocabSize]

        return logits;
    }

    // Tornar as propriedades maxSeqLen e vocabSize acessíveis publicamente
     public int MaxSeqLen => maxSeqLen;
     public int VocabSize => vocabSize;


    public void Save(string path)
    {
        try
        {
            // CORREÇÃO: Gerenciar a memória dos tensores NO state_dict com DisposeScope
            // A declaração 'using' aplica-se ao DisposeScope, NÃO ao stateDict
            this.save(path);

            Console.WriteLine($"Modelo salvo (usando Module.save) em: {path}");

            // O scope.Dispose() será chamado automaticamente aqui ao sair do using block,
            // liberando os tensores que foram registrados quando state_dict() foi chamado.
        }
        catch (Exception ex)
        {
            Console.WriteLine($"ERRO ao salvar o modelo em {path}: {ex.Message}");
            Console.WriteLine(ex.StackTrace);
        }
    }

    public void Load(string path)
    {
        if (!File.Exists(path))
        {
             Console.WriteLine($"AVISO: Arquivo de modelo não encontrado em {path}. Não foi possível carregar.");
             return;
        }

        try
        {
            device = TorchSharp.torch.cuda.is_available() ? CUDA : CPU;
            this.to(device); // Mover modelo para o device ANTES de carregar

            // CORREÇÃO: Usar o método Load do próprio Module
            // Este método carrega o estado COMPLETO do módulo a partir de um arquivo salvo por Module.save
            // Verificar se na sua versão (0.105.0) o método load tem um parâmetro 'strict'.
            // Se o erro "Cannot resolve symbol 'strict'" ocorrer aqui, remova ', strict: false'.
            this.load(path, strict: false); // Tentar com strict: false primeiro

            Console.WriteLine($"Modelo carregado (usando Module.load) de {path}");
        }
        catch (Exception ex)
        {
             Console.WriteLine($"ERRO geral ao carregar modelo de {path}: {ex.Message}");
             Console.WriteLine(ex.StackTrace);
             // Considere lançar a exceção ou ter um comportamento de fallback
             // throw;
        }
    }
}
// --- END OF FILE TransformerModel.cs ---

File: AI/LLM/core/transformer_model.py
# core/transformer_model.py

import torch
import torch.nn as nn # Importar nn
# Importar outras classes que TransformerModel usa
from .transformer_block import TransformerBlock
# from .feed_forward import FeedForward # Importado via TransformerBlock
# from .multi_head_self_attention import MultiHeadSelfAttention # Importado via TransformerBlock
from typing import List, Tuple # Importar tipos
import os # Para lidar com caminhos de arquivo e diretórios
# import math # Importado via MultiHeadSelfAttention
# import sys # Importado via MultiHeadSelfAttention

class TransformerModel(nn.Module):
    """
    Modelo completo Transformer (simplificado).
    """
    def __init__(self, vocab_size: int, max_seq_len: int = 512,
                 embedding_dim: int = 256, num_heads: int = 4,
                 num_layers: int = 2, dropout: float = 0.1):
        # --- PASSO 1: CHAMAR O CONSTRUTOR DA CLASSE BASE (SEMPRE O PRIMEIRO) ---
        super().__init__() # ESSENCIAL e a PRIMEIRA linha!

        # Determina o dispositivo
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        print(f"Modelo será executado em: {self.device}") # Manter print de device


        # Salvar parâmetros principais como atributos
        self.vocab_size = vocab_size
        self.max_seq_len = max_seq_len
        self.embedding_dim = embedding_dim
        self.dropout_prob = dropout # Salvar a probabilidade de dropout


        # --- PASSO 2: INSTANCIAR SUB-MÓDULOS E DEFINIR ATRIBUTOS ---
        self.token_embedding = nn.Embedding(vocab_size, embedding_dim)
        self.position_embedding = nn.Embedding(max_seq_len, embedding_dim) # Embedding posicional aprendível

        self.transformer_blocks = nn.ModuleList([
            TransformerBlock(embedding_dim, num_heads, max_seq_len, dropout) # Passar todos os args
            for _ in range(num_layers)
        ])

        self.final_norm = nn.LayerNorm(embedding_dim) # Normalização final
        self.output_linear = nn.Linear(embedding_dim, vocab_size) # Camada de saída


        # --- PASSO 3: APLICAR INICIALIZAÇÕES OU CONFIGURAÇÕES GLOBAIS ---
        self.apply(self._init_weights) # Aplicar inicialização de pesos

        self.to(self.device) # Mover o modelo completo para o dispositivo


    # --- DEFINIR O MÉTODO _init_weights NA CLASSE ---
    def _init_weights(self, module):
       """
       Inicializa os pesos de módulos específicos.
       Aplicado recursivamente a todos os sub-módulos.
       """
       if isinstance(module, nn.Linear):
           torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
           if module.bias is not None:
               torch.nn.init.zeros_(module.bias)
       elif isinstance(module, nn.Embedding):
           torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
       elif isinstance(module, nn.LayerNorm):
           if hasattr(module, 'bias') and module.bias is not None:
               torch.nn.init.zeros_(module.bias)
           if hasattr(module, 'weight') and module.weight is not None:
                torch.nn.init.ones_(module.weight)

    # --- DEFINIR O MÉTODO forward ---
    def forward(self, input_ids: torch.Tensor) -> torch.Tensor:
        # input_ids shape: (B, T_input) - Batch, Input Sequence Length
        # --- ESSENCIAIS PRIMEIRO ---
        B, T_input = input_ids.shape
        # ---------------------------

        # O forward do modelo espera uma sequência de tamanho ATÉ max_seq_len.
        # Se a sequência de entrada for maior, precisamos TRUNCAR aqui.
        if T_input > self.max_seq_len:
            # Pega APENAS os últimos max_seq_len tokens
            input_ids = input_ids[:, -self.max_seq_len:]
            T_input = self.max_seq_len # Atualiza T_input


        # Criar tensor de posições dinamicamente NO MESMO DISPOSITIVO DO INPUT
        positions = torch.arange(0, T_input, dtype=torch.long, device=input_ids.device) # (T_input,)
        positions = positions.unsqueeze(0).expand(B, T_input) # (B, T_input)


        # Calcular embeddings + posições
        token_embeds = self.token_embedding(input_ids) # [B, T_input, embed_dim]
        pos_embeds = self.position_embedding(positions) # [B, T_input, embed_dim]
        x = token_embeds + pos_embeds # [B, T_input, embed_dim]

        # Aplicar dropout no embedding (opcional)
        # x = torch.nn.functional.dropout(x, p=self.dropout_prob, training=self.training)

        # Passar pelos blocos Transformer
        for block in self.transformer_blocks: # Itera sobre os módulos em ModuleList
            x = block(x) # Chamada forward implícita

        # Normalização final
        x = self.final_norm(x)

        # Camada linear final para obter logits sobre o vocabulário
        logits = self.output_linear(x) # [B, T_input, vocab_size]

        return logits # Retornar o tensor de logits

    # --- DEFINIR PROPRIEDADES ---
    @property
    def MaxSeqLen(self) -> int:
        return self.max_seq_len

    @property
    def VocabSize(self) -> int:
        return self.vocab_size


    # --- DEFINIR MÉTODOS save/load ---
    def save(self, path: str):
        """
        Salva o estado do modelo (pesos) em um arquivo.
        """
        try:
            save_dir = os.path.dirname(path)
            if save_dir and not os.path.exists(save_dir):
                os.makedirs(save_dir, exist_ok=True)
                print(f"Diretório criado para salvar modelo: {save_dir}")

            torch.save(self.state_dict(), path)
            print(f"Estado do modelo salvo em: {path}")
        except Exception as e:
            print(f"ERRO ao salvar o modelo em {path}: {e}")
            # import traceback; traceback.print_exc()


    def load(self, path: str):
        """
        Carrega o estado do modelo (pesos) de um arquivo.
        """
        if not os.path.exists(path):
             # print(f"AVISO: Arquivo de modelo não encontrado em {path}. Não foi possível carregar.") # Este print já está na lógica condicional em app.py
             return # Sai da função se o arquivo não existir

        try:
            target_device = self.device
            state_dict = torch.load(path, map_location=target_device)
            self.load_state_dict(state_dict, strict=False)

            print(f"Estado do modelo carregado de {path}")
        except Exception as e:
             print(f"ERRO geral ao carregar modelo de {path}: {e}")
             # import traceback; traceback.print_exc()
             # raise e # Opcional: re-lançar se o carregamento for crítico